### DISCUSSION TITLE
[Proposal]: Singleton object syntax

### DISCUSSION THREAD

**Discussion by Cywohoy** [dotnet/csharplang#5610] [Language Ideas] [answered]
# Singleton object syntax

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
Singleton class like kotlin's `object`

## Motivation
[motivation]: #motivation

When we use a singleton class, there is a problem of using an `Instance` when calling each member. This makes the code longer and looks not clear. 

## Detailed design
[design]: #detailed-design

New singleton object using `object`
```cs
public object class A {
    public string a;
    public void Test() {
        System.Console.WriteLine($"a: {a}");
    }

    //singleton's constructor similar to static constructor
    A() {
        a = "hello world";
    }

    //static members not allowed
    //public static int b;
    //public static void TestStatic() {...}
}

A.Test();
// a: hello world
```
can also inherit other class, or implement interfaces.
```cs
public class Base {
    public int a;
    public int b;

    public virtual void Test() {
        System.Console.WriteLine($"a + b = {a+b}");
    } 
}

public object Singleton : Base {
    public int c;
    public override void Test() {
        System.Console.WriteLine($"a + b + c = {a+b+c}");
    }
}

Base a = Singleton;
a.a = 1;
a.b = 2;
Singleton.c = 3;

a.Test();
// a + b + c = 6
```
maybe generic singleton can be implemented
```cs
public object DefaultValue<T> where T : struct {
    public T Value;
    
    DefaultValue() {
        Value = default(T);
    }
}

System.Console.WriteLine(DefaultValue<bool>.Value)
// false
```
## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

general singleton classes, and static classes.

## Unresolved questions
[unresolved]: #unresolved-questions

`object struct`.

**Comment by DavidArno** [comment]
Under "drawbacks", you forgot to list the fact that most people regard the singleton pattern as the archetypal anti-pattern as it creates global state which in turn makes testing far harder than it needs to be. Encapsulating an anti-pattern into the C# language would be a huge mistake.

**Reply by YairHalberstadt** [reply]
A singleton does not necessarily encapsulate state - see for example EqualityComparer.Default which is stateless. This is rather common, to have a stateless singleton object which implements an interface.

**Reply by DavidArno** [reply]
@YairHalberstadt,

Indeed, value/deterministic singletons are very common. They are also unrelated to the infamous singleton pattern.

**Reply by YairHalberstadt** [reply]
This particular feature would be useful for stateless singletons too - scala has this and it's commonly used for that purpose.

**Reply by TahirAhmadov** [reply]
Yes the comparer singleton is a valid use case but there are literally a few types like this if you look at the entirety of BCL and all code. Why add a feature to the language which has one good use case, and a dozen really bad ones? This is an argument I had a few times with my colleagues and one I didn't always completely win - people stick these singletons everywhere; making it easier or more natural will make this much worse.
Secondly, this goes against the core language structure: types can have static members, but not instance members. With this proposed syntax, it's unclear whether the identifier is a type or a singleton variable of some sort.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/490
